### ISSUE TITLE
TF does not refresh resource state in each apply step for resources that have implicit dependencies

### ISSUE THREAD

**Issue by ravitandonrt** [hashicorp/terraform-plugin-sdk#121]
Looking for advice and/or possible enhancement in Terraform on how to handle deletion of implicit dependent resources, where the primary resource is getting deleted and created via `taint` command while the dependencies (also expressed in terraform config) are removed during execution by the service, leading to discrepancy in the terraform state and apply to fail. 

Using depends_on or an interpolation does not work as the changing field is optional and does not warrant a forceNew on the dependent resource, hence an update is invoked on a deleted resource.

- Are changes happening in 0.12 or later versions with regards to dependency walk for tainted resources going to address this? Can taint and depends_on work together?
- Is having a 'fake' field with forceNew in the such implicitly deletable resources a viable alternative?
- Calling state refresh/get before each update seem like an overkill. Right now refresh only happens at start of apply plan, but in this case changes happen during the execution. It will definitely complicate provider logic to now decide that an update needs to be a create instead and maybe at odds with Terraform behavior.

### Terraform Version

```
0.11.8
```

### Terraform Configuration Files

```hcl
provider "oci" {
  tenancy_ocid      = "${var.tenancy_ocid}"
  user_ocid         = "${var.user_ocid}"
  private_key_path  = "${var.private_key_path}"
  fingerprint       = "${var.fingerprint}"
  region            = "${var.region}"
}

resource "oci_core_instance" "instance" {
  compartment_id = "${var.compartment_ocid}"
  availability_domain = "${lookup(data.oci_identity_availability_domains.ADs.availability_domains[var.availability_domain - 1],"name")}"

  shape = "${var.shape}"
  source_details {
    source_type = "image"
    source_id = "${var.image_ocid}"
  }

  create_vnic_details {
    subnet_id = "${oci_core_subnet.ExampleSubnet.id}"
    assign_public_ip = "false"
  }
}

data "oci_core_private_ips" "primary_private_ip" {
  subnet_id = "${oci_core_instance.instance.subnet_id}"
  ip_address = "${oci_core_instance.instance.private_ip}"
}

resource "oci_core_private_ip" "secondary" {
  #vnic_id = "${lookup(data.oci_core_private_ips.primary_private_ip.private_ips[0], "vnic_id")}"
  vnic_id = "${data.oci_core_private_ips.primary_private_ip.private_ips.0.vnic_id}"
}
```

Full repro with networking template and full logs is also available in this gist link: https://gist.github.com/ravitandonrt/4c141a728b980087aa065f687b3a8be3

### Debug Output

Refer https://gist.github.com/ravitandonrt/4c141a728b980087aa065f687b3a8be3
The gist contains terraform-output.log that has detailed log of the second apply (post taint) where this happens. First two steps have no issues.

### Crash Output

No crashes, only apply failure as noted above.

### Expected Behavior

The apply should have succeeded in first try post the taint, that is, Terraform/Provider should have detected the resource got deleted during the apply steps and initiate a create instead of an update, or, provide a way for taint to understand depends_on so that those resources are automatically tainted.

### Actual Behavior

The deletion of an OCI instance causes (service behavior) the vnics and any secondary private_ip resources attached to it be implicitly deleted, however the secondary private_ip resource continues to exist in Terraform state, causing Terraform to process an update on a deleted resource.

Right now, the work-arounds are either to taint all dependencies separately or use a destroy with target (which actually deletes the dependencies) or do double applies.

### Steps to Reproduce

`terraform apply`
`terraform taint oci_core_instance.instance`
`terraform apply`

### Additional Context

Yes, Terraform is running in a CI system, so doing double applies is not ideal.

Customer cannot use create_before_destroy (not that it matters) as these instances need to use static ip addresses.
 
### References

https://github.com/terraform-providers/terraform-provider-oci/issues/613

**Comment by paultyng** [comment]
Are you able to repro this in 0.12? 0.12.16 and 0.12.14 had fixes around destroy and taint scenarios.

**Comment by radeksimko** [comment]
Hi @ravitandonrt 

> Are changes happening in 0.12 or later versions with regards to dependency walk for tainted resources going to address this? Can taint and depends_on work together?

In the context you described I believe that `taint` & `depends_on` continue to work together the same way. I believe that the (mis)behaviour you're describing can be triggered by changing any field that has `ForceNew` on it, so it's not directly related to tainting. Also it's important to note that `depends_on` creates context-less resource to resource relationship, unlike referencing via fields.

Also `depends_on` is meant to be used as the last course of action when referencing isn't possible. While it is a useful tool we'd like to limit the number of reasons users need to use it for as it increases the overall complexity of configuration and hurts readability.

> Is having a 'fake' field with forceNew in the such implicitly deletable resources a viable alternative?

I think the best course of action here would be SDK exposing a way for providers to define implicit dependencies, so that Terraform can build the graph in such a way that resources which (implicitly) depend on another are scheduled for recreation too and the recreation of these is scheduled before recreation of the resource being tainted/modified.

> Calling state refresh/get before each update seem like an overkill. Right now refresh only happens at start of apply plan, but in this case changes happen during the execution. It will definitely complicate provider logic to now decide that an update needs to be a create instead and maybe at odds with Terraform behavior.

I think in ideal case Terraform just needs to get the plan right so there is no need for further refresh. For that to happen though it may need some hinting from providers as only providers know how certain changes affect certain fields or resources.

Accurate planning and predictability is one of the key features Terraform users rely on, so that's the path we'd likely prefer over refreshing resources and data sources downstream in the graph after their ancestors were recreated or modified. That said https://github.com/hashicorp/terraform/issues/4149 describes how we might address cases where it's impossible to predict changes.

---

I believe this is a duplicate of https://github.com/hashicorp/terraform-plugin-sdk/issues/122 - the example there just uses GCP's storage object & object ACL which also have an implicit relationship impossible to define in the schema today.

Do you mind me closing this issue in favour of #122 so we can keep discussion in one place?

**Comment by alexng-canuck** [comment]
Hi @radeksimko,

I'm a colleague of Ravi's and he is no longer working on this issue. Having read your description of the issue, I think it makes sense to close this and to track any fixes as part of #122 .

Thanks,
Alex
